File: USACO/2023-2024/Feb 2024/Gold/quantummoochanics.java
import java.util.*;
import java.io.*;

public class quantummoochanics {
	static long[] p, s;
	public static void main(String[] args) throws IOException {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		int T = Integer.parseInt(st.nextToken());
		for(int t = 0; t < T; t++) {
			st = new StringTokenizer(br.readLine());
			int N = Integer.parseInt(st.nextToken());
			int[][] adj = new int[N][2];
			p = new long[N];
			s = new long[N];
			st = new StringTokenizer(br.readLine());
			for(int i = 0; i < N; i++) p[i] = Long.parseLong(st.nextToken());
			st = new StringTokenizer(br.readLine());
			for(int i = 0; i < N; i++) s[i] = Integer.parseInt(st.nextToken());
			
			long[] res = new long[N];
			PriorityQueue<Pair> toVisit = new PriorityQueue<>();
			for(int i = 0; i < N; i++) {
				adj[i] = new int[]{i-1, i+1};
				if(i < N-1) toVisit.add(new Pair(i, i+1));
			}
			
			while(!toVisit.isEmpty()) {
				Pair curr = toVisit.remove();
				if(res[curr.a] != 0 || res[curr.b] != 0) continue;
				res[curr.a] = res[curr.b] = curr.obsnum;
				int l = adj[curr.a][0], r = adj[curr.b][1];
				if(l >= 0 && r < N) {
					adj[l][1] = r;
					adj[r][0] = l;
					toVisit.add(new Pair(l, r));
				}
			}
			
			for(int i = 0; i < N; i++) {
				System.out.print(res[i]);
				if(i < N-1) System.out.print(" ");
			}
			System.out.println();
		}
		br.close();
	}
	
	static class Pair implements Comparable<Pair> {
		int a, b;
		long obsnum;
		
		Pair(int a, int b) {
			this.a = a;
			this.b = b;
			this.obsnum = (long) Math.ceil((double)(p[b] - p[a]) / (s[a] + s[b]));
			this.obsnum = this.obsnum * 2 - (a % 2 == 0 ? 1 : 0);
		}
		
		public int compareTo(Pair other) {
			return Long.compare(obsnum, other.obsnum);
		}
	}
}

File: USACO/2023-2024/Feb 2024/Gold/quantummoochanics.cpp
#include <bits/stdc++.h>

#define ll long long
#define f first
#define s second

using namespace std;

ll p[200000], s[200000], res[200000];

ll calc_num(pair<int, int> a) {
	ll speed_sum = s[a.f] + s[a.s];
	return ((p[a.s]-p[a.f]+speed_sum-1)/speed_sum)*2 - (a.f%2 == 0);
}

int main() {
	int T; cin >> T;
	while(T--) {
		int N; cin >> N;
		for(int i = 0; i < N; i++) cin >> p[i];
		for(int i = 0; i < N; i++) cin >> s[i];
		
		priority_queue<pair<ll, pair<int, int>>> pq;
		
		for(int i = 0; i < N; i++) res[i] = 0;
		pair<int, int> adj[N];
		for(int i = 0; i < N; i++) {
			adj[i] = {i-1, i+1};
			if(i < N-1) pq.push({-calc_num({i, i+1}), {i, i+1}});
		}
		
		while(pq.size()) {
			ll num = -pq.top().f;
			pair curr = pq.top().s;
			pq.pop();
			if(res[curr.f] || res[curr.s]) continue;
			
			res[curr.f] = res[curr.s] = num;
			int l = adj[curr.f].f, r = adj[curr.s].s;
			if(l >= 0 && r < N) {
				adj[l].s = r;
				adj[r].f = l;
				pq.push({-calc_num({l, r}), {l, r}});
			}
		}
		
		for(int i = 0; i < N; i++) cout << res[i] << " \n"[i == N - 1];
	}
}
